8337511: Implement JEP-404: Generational Shenandoah (Experimental)

earthling-amzn opened the pull request:
This PR merges JEP 404, a generational mode for the Shenandoah garbage collector. The JEP can be viewed here: https://openjdk.org/jeps/404. We would like to target JDK24 with this PR.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [ ] Change requires a JEP request to be targeted

### Issues
 * [JDK-8337511](https://bugs.openjdk.org/browse/JDK-8337511): Implement JEP-404: Generational Shenandoah (Experimental) (**Sub-task** - P4)
 * [JDK-8260865](https://bugs.openjdk.org/browse/JDK-8260865): JEP 404: Generational Shenandoah (Experimental) (**JEP**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20395/head:pull/20395` \
`$ git checkout pull/20395`

Update a local copy of the PR: \
`$ git checkout pull/20395` \
`$ git pull https://git.openjdk.org/jdk.git pull/20395/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20395`

View PR using the GUI difftool: \
`$ git pr show -t 20395`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20395.diff">https://git.openjdk.org/jdk/pull/20395.diff</a>

</details>

earthling-amzn commented:
/jep JEP-404

rkennke commented on src/hotspot/cpu/ppc/gc/shenandoah/c1/shenandoahBarrierSetC1_ppc.cpp:
```diff
@@ -102,11 +102,21 @@ LIR_Opr ShenandoahBarrierSetC1::atomic_cmpxchg_at_resolved(LIRAccess &access, LI
-  return BarrierSetC1::atomic_cmpxchg_at_resolved(access, cmp_value, new_value);
+  LIR_Opr result = BarrierSetC1::atomic_cmpxchg_at_resolved(access, cmp_value, new_value);
+
+  if (ShenandoahCardBarrier && access.is_oop()) {
```
Same problem as in aarch64 path.

earthling-amzn replied:
I agree, will remove this.

rkennke commented on src/hotspot/cpu/riscv/gc/shenandoah/shenandoahBarrierSetAssembler_riscv.cpp:
```diff
@@ -64,7 +65,7 @@ void ShenandoahBarrierSetAssembler::arraycopy_prologue(MacroAssembler* masm, Dec
         __ test_bit(t0, t0, ShenandoahHeap::HAS_FORWARDED_BITPOS);
         __ beqz(t0, done);
       } else {
-        __ andi(t0, t0, ShenandoahHeap::HAS_FORWARDED | ShenandoahHeap::MARKING);
```
I don't see corresponding changes in the other arches. OTOH, a lot of changes that are in other arches are missing in riscv. Maybe drop the changes and make a separate and complete effort to implement riscv?

earthling-amzn replied:
Yes, it looks the RiscV port is missing quite a lot of the generational mode changes. I'll file a ticket to make the generational mode unavailable on this platform and backout the rest of the RiscV changes.

earthling-amzn replied:
Actually, we do block the generational mode for RiscV. Some of these changes are necessary because of changes to shared symbols. I'll take a closer look.

earthling-amzn replied:
These changes are harmless, but unnecessary after JDK-8314777. Will revert.

rkennke commented on src/hotspot/cpu/x86/gc/shenandoah/c1/shenandoahBarrierSetC1_x86.cpp:
```diff
@@ -84,10 +85,21 @@ LIR_Opr ShenandoahBarrierSetC1::atomic_cmpxchg_at_resolved(LIRAccess& access, LI
+
+  LIR_Opr result = BarrierSetC1::atomic_cmpxchg_at_resolved(access, cmp_value, new_value);
+
+  if (ShenandoahCardBarrier && access.is_oop()) {
```
Same issue as in other arches.

rkennke commented on src/hotspot/cpu/x86/gc/shenandoah/shenandoahBarrierSetAssembler_x86.cpp:
```diff
@@ -140,10 +165,10 @@ void ShenandoahBarrierSetAssembler::arraycopy_prologue(MacroAssembler* masm, Dec
       assert_different_registers(src, dst, count, thread);
 
-      Label done;
+      Label L_done;
```
No need to rename the label? (This only clutters the patch and serves no purpose, IMO)

earthling-amzn replied:
This convention is used in many places in the barrier code:
```
% grep -riP 'Label.* L_done' src/hotspot | wc -l
44
```

rkennke commented on src/hotspot/share/gc/shared/gcConfiguration.cpp:
```diff
@@ -66,18 +71,22 @@ GCName GCConfiguration::old_collector() const {
+      return Shenandoah;
+    }
+#endif
+    return NA;
```
This looks like a changed behaviour: when generational mode is not selected, we now return NA, while before we returned Shenandoah. We should probably keep returning Shenandoah there.

earthling-amzn replied:
Agree. Will fix.

rkennke reviewed (changes requested):
Alright, shared non-Shenandoah changes are *really* few, but I review them separately. I think I spotted a change of behaviour (in Shenandoah path, but in shared code).

rkennke reviewed (changes requested):
And here comes the first batch of Shenandoah-shared-changes review.

jaskarth commented on test/hotspot/jtreg/compiler/c2/irTests/TestIfMinMax.java:
```diff
@@ -33,7 +33,6 @@
  * @bug 8324655 8329797
  * @key randomness
  * @summary Test that if expressions are properly folded into min/max nodes
- * @requires os.arch != "riscv64"
```
This is just a drive-by but I think this change reverts [JDK-8329355](https://bugs.openjdk.org/browse/JDK-8329355). It might have been accidentally removed in this merge: https://github.com/openjdk/jdk/commit/e30fa929b19e8b7d6db9073fdf640938d3db39fd

earthling-amzn replied:
Good catch. I sync'd this file back up with `openjdk/jdk:master`.

ysramakrishna commented on src/hotspot/share/gc/shenandoah/shenandoahCollectionSet.hpp:
```diff
@@ -43,10 +52,28 @@ class ShenandoahCollectionSet : public CHeapObj<mtGC> {
+  // Points to array identifying which tenure-age regions have been preselected
+  // for inclusion in collection set. This field is only valid during brief
+  // spans of time while collection set is being constructed.
+  bool*                 _preselected_regions;
```
I agree that this code is a bit smelly. As you note, it has a very awkward life-cycle and ideally would be an automatic variable that is freed when its usage scope is exited. At this time, it's resource-area allocated and freed when we exit the resource mark within which it was allocated. (Originally, it used to be `alloca`ed on the stack, which itself was problematic.)

The main role appears to be to attach a predicate called "preselected" to associate with regions that have already been included in a collection set based on an earlier pass which can therefore be ignored by later passes that are adding to the collection set.

As it's entirely associated with collection set construction, it seemed reasonable to place the field in the collection set object.

The alternative is to pass the region-indexed predicate array around between the methods that use them, or put it in the heuristics object. I agree that the structure of this code could do with some refactoring to clean this up some more.

rkennke replied:
Ok, let's investigate this later.

ysramakrishna replied:
Created https://bugs.openjdk.org/browse/JDK-8338472 to track this work.
Resolving comment thread.

rkennke reviewed (changes requested):
I'm finishing my day and send out today's round of reviews.
I must say that in many places I find it difficult to convince myself that the changes in shared Shenandoah code don't affect singly-gen-Shenandoah. But at the same time, I am not sure what we can do about it. We certainly don't want to spend another year refactoring this, and risk to introduce new bugs. I suspect that you guys did a lot of due diligence to get this correct. I'd welcome your perspective on this matter.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/heuristics/shenandoahAdaptiveHeuristics.cpp:
```diff
@@ -21,16 +22,21 @@
 #include "gc/shenandoah/shenandoahHeapRegion.inline.hpp"
 #include "logging/log.hpp"
 #include "logging/logTag.hpp"
+#include "runtime/globals_extension.hpp"
```
Yes, just `globals.hpp` here.

RealFYang commented:
Hi, I have prepared needed changes for riscv by adapting from aarch64 changes.
This has passed `gtest:all` and `hotspot_gc_shenandoah` on linux-riscv64 with both fastdebug and release builds.

[20395-riscv.diff.txt](https://github.com/user-attachments/files/16490335/20395-riscv.diff.txt)

ysramakrishna reviewed (commented):
Flushing some accumulated comments. Will pick up aagain.

rkennke commented on src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:
```diff
@@ -434,19 +473,29 @@ jint ShenandoahHeap::initialize() {
+  initialize_controller();
 
-  ShenandoahInitLogger::print();
+  print_init_logger();
```
Is it worth to replace the one-liners with multi-line method calls?

rkennke commented on src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:
```diff
@@ -683,33 +684,84 @@ void ShenandoahHeap::decrease_committed(size_t bytes) {
-    if (waste) {
-      pacer()->claim_for_alloc(words, true);
+    if (waste > 0) {
+      pacer()->claim_for_alloc(waste, true);
```
Is passing 'waste' to the pacer correct here?

earthling-amzn replied:
Yes, previously, if `waste` was `true` it meant that `words` was wasted words from the allocation (due to alignment or not fitting in the region). Now, `waste` is just that the value (rather than a flag indicating another variable holds the value).

rkennke reviewed (changes requested):
Some more comments that I need to flush out...

rkennke commented on src/hotspot/share/gc/shenandoah/shenandoahHeap.hpp:
```diff
@@ -666,7 +781,16 @@ class ShenandoahHeap : public CollectedHeap, public ShenandoahSpaceInfo {
+  // Return the object's age, or a sentinel value when the age can't
+  // necessarily be determined because of concurrent locking by the
+  // mutator
+  static inline uint get_object_age(oop obj);
```
Once https://github.com/openjdk/jdk/pull/20067 is integrated, you could force LW-locking with that new monitor-mapping and enjoy simple hassle-free updating of the age bits in the object header.

earthling-amzn replied:
@ysramakrishna will create a tracking ticket so that we don't forget about this.

rkennke commented on src/hotspot/share/gc/shenandoah/shenandoahHeap.inline.hpp:
```diff
@@ -302,6 +454,7 @@ inline bool ShenandoahHeap::in_collection_set_loc(void* p) const {
   return collection_set()->is_in_loc(p);
 }
 
+
```
Excessive newline.

earthling-amzn replied:
Yep.

rkennke reviewed (changes requested):
Some more reviews, it should complete shenandoahHeap.* files.

ysramakrishna commented on src/hotspot/share/gc/shenandoah/shenandoahHeap.hpp:
```diff
@@ -136,11 +157,46 @@ class ShenandoahHeap : public CollectedHeap, public ShenandoahSpaceInfo {
+  // seen by all mutators as well as all GC worker threads.
+  // In that sense, it's a stable snapshot of _gc_generation that is
+  // updated at each STW pause associated with a ShenandoahVMOp.
+  ShenandoahGeneration* _active_generation;
```
Not as currently set up. The controller thread currently changes the first asynchronously (wrt mutators). The shadow copy in _active_generation is now used to buffer against this asynchrony. See https://github.com/openjdk/shenandoah/pull/407 for the changes and a description.

rkennke replied:
Ok, but it looks as if _gc_generation should just be set at a safepoint, then. Or is this not possible?

rkennke commented on src/hotspot/share/gc/shenandoah/shenandoahMark.inline.hpp:
```diff
@@ -58,9 +62,13 @@ void ShenandoahMark::dedup_string(oop obj, StringDedup::Requests* const req) {
 
+  // TODO: This will push array chunks into the mark queue with no regard for
+  // generations. I don't think it will break anything, but the young generation
+  // scan might end up processing some old generation array chunks.
```
Shouldn't this code push to the same queue from which it originally popped the last task? AFAIU, chunks can only cross generation boundaries if a worker actively steals it from another queue, and I don't know if that is even possible between generations.

earthling-amzn replied:
@earthling-amzn will delete this on his `TODO` clean up PR.

rkennke commented on src/hotspot/share/gc/shenandoah/shenandoahMark.inline.hpp:
```diff
@@ -104,14 +112,27 @@ void ShenandoahMark::do_task(ShenandoahObjToScanQueue* q, T* cl, ShenandoahLiveD
+  const size_t size = obj->size();
+
+  // Age census for objects in the young generation
+  if (GENERATION == YOUNG || (GENERATION == GLOBAL && region->is_young())) {
```
If I read this correctly, this condition will generate a runtime check region->is_young() in single-gen Shenandoah. 1. This seems like something that would never happen (in single-gen GC) and 2. This path is critical for performance of concurrent marking. Please make sure to keep it as fast as possible, also with respect to the following flag-checks.

earthling-amzn replied:
There is a separate `ShenandoahGenerationType` for non-generational modes: `NON_GEN`. https://github.com/openjdk/jdk/pull/20395/files/46638cdef26fd25c290e6f63035ff5d2f1dc3bdb#diff-87bed125e3733038ed0310afc884fed6623fe7abe4f04bf7b8e0fd45ab73e79c

rkennke replied:
ok, then.

rkennke commented on src/hotspot/share/gc/shenandoah/shenandoahMemoryPool.hpp:
```diff
@@ -25,21 +26,46 @@
 #ifndef SHARE_GC_SHENANDOAH_SHENANDOAHMEMORYPOOL_HPP
 #define SHARE_GC_SHENANDOAH_SHENANDOAHMEMORYPOOL_HPP
 
-#ifndef SERIALGC
```
Huh, I wonder what that ifndef was for. Apparently it was there since JEP189 appeared. Especially since SERIALGC is practically always defined - how did the includes even work? *scratches head*

earthling-amzn replied:
'Tis strange, to be sure.

rkennke reviewed (changes requested):
More review comments.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.cpp:
```diff
@@ -100,11 +117,35 @@ void ShenandoahHeapRegion::make_regular_allocation() {
+// behavior previously performed as a side effect of make_regular_bypass().  This is used by Full GC
+void ShenandoahHeapRegion::make_young_maybe() {
+  shenandoah_assert_heaplocked();
+  assert(!ShenandoahHeap::heap()->mode()->is_generational(), "Only call if non-generational");
```
Agree, this method name is confusing. Regions can be affiliated with the young or old generation - or, they can also be unaffiliated. When a region is unaffiliated, its affiliation field is set to `FREE`. In non-generational modes, all regions are either `FREE` or affiliated with the "young" generation - which is essentially just "not-free" in non-generational modes. Perhaps we should call this method `make_affiliated_maybe` instead?

rkennke replied:
That would be better, yes. And add a comment explaining it.

earthling-amzn replied:
https://bugs.openjdk.org/browse/JDK-8338528

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp:
```diff
@@ -671,109 +840,249 @@ HeapWord* ShenandoahFreeSet::allocate_single(ShenandoahAllocRequest& req, bool&
+      // only for old-gen evacuations.
+
+      // TODO:
+      // if (GC is idle (out of cycle) and mutator allocation fails and there is memory reserved in Collector
```
I'm working on cleaning up all of our lingering `TODOs`.

earthling-amzn commented:
> Hi, I have prepared needed changes for riscv by adapting from aarch64 changes. This has passed `gtest:all` and `hotspot_gc_shenandoah` on linux-riscv64 with both fastdebug and release builds.
> 
> [20395-riscv.diff.txt](https://github.com/user-attachments/files/16490335/20395-riscv.diff.txt)

@RealFYang - could you open a JBS ticket for riscv support for genshen and open a PR against the master branch in this repo: https://github.com/openjdk/shenandoah. Thank you!

earthling-amzn commented on src/hotspot/share/gc/shenandoah/mode/shenandoahSATBMode.cpp:
```diff
@@ -48,22 +44,6 @@ void ShenandoahSATBMode::initialize_flags() const {
-  }
-  vm_exit_during_initialization("Unknown -XX:ShenandoahGCHeuristics option");
-  return nullptr;
+  assert(strcmp(ShenandoahGCMode, "generational") != 0, "Error");
```
I agree - we have asserts that test for generational mode. I'll make a complimentary assert for non-generational mode.

rkennke commented on src/hotspot/share/gc/shenandoah/shenandoahNMethod.cpp:
```diff
@@ -126,13 +126,13 @@ void ShenandoahNMethod::heal_nmethod(nmethod* nm) {
     heal_nmethod_metadata(data);
+  } else if (heap->is_concurrent_mark_in_progress()) {
+    ShenandoahKeepAliveClosure cl;
+    data->oops_do(&cl);
```
AFAICT, all that this change does is swap the condition ordering. Is this significant in any way?

earthling-amzn replied:
Yes, we did this so the evacuation phase would take precedence over concurrent marking of the old generation. That is, it is, when old marking is happening, it may also be the case that `is_concurrent_weak_root_in_progress` or `is_concurrent_strong_root_in_progress` and `is_concurrent_mark_in_progress` may all be true. However, the generational mode's old collection doesn't directly evacuate, so when roots are in progress, we want the evacuation case.

rkennke commented on src/hotspot/share/gc/shenandoah/shenandoahStackWatermark.cpp:
```diff
@@ -73,11 +74,11 @@ OopClosure* ShenandoahStackWatermark::closure_from_context(void* context) {
       assert(_heap->is_evacuation_in_progress(), "Nothing to evacuate");
       return &_evac_update_oop_cl;
+    } else if (_heap->is_concurrent_mark_in_progress()) {
+      return &_keep_alive_cl;
```
Swapping conditions, again. Can concurrent mark overlap with concurrent roots, now, and this is why this is relevant?

earthling-amzn replied:
See https://github.com/openjdk/jdk/pull/20395#discussion_r1720473711

rkennke commented on src/hotspot/share/gc/shenandoah/shenandoahUnload.cpp:
```diff
@@ -51,7 +51,8 @@ class ShenandoahIsUnloadingOopClosure : public OopClosure {
   ShenandoahIsUnloadingOopClosure() :
-    _marking_context(ShenandoahHeap::heap()->complete_marking_context()),
+    // TODO: In non-generational mode, this should still be complete_marking_context()
+    _marking_context(ShenandoahHeap::heap()->marking_context()),
```
Is this not causing troubles in non-generational mode?

earthling-amzn replied:
`complete_marking_context` just asserts that marking is complete before returning the mark context. I'll revert this and test.

earthling-amzn replied:
https://bugs.openjdk.org/browse/JDK-8339127

rkennke reviewed (changes requested):
Some more reviews. I am now done (first-pass) with changes in existing files in src/hotspot/share/gc/shenandoah.

rkennke commented on test/hotspot/jtreg/gc/stress/jfr/TestStressBigAllocationGCEventsWithShenandoah.java:
```diff
@@ -0,0 +1,50 @@
+ * @test id=default
+ * @key randomness
+ * @requires vm.hasJFR
+ * @requires vm.gc == "Shenandoah"
```
Use '@requires vm.gc.Shenandoah' instead. The semantics are somewhat different: '@requires vm.gc == "Shenandoah"' means 'Shenandoah has been selected at the command line when running jtreg' (which kinda implies that OpenJDK has been built with Shenandoah), which also means that this would have no effect really, while '@requires vm.gc.Shenandoah' means 'OpenJDK under test has been built with Shenandoah GC', which is really what we want: if somebody (e.g. Oracle) builds the JDK without Shenandoah, then we want to skip this test configuration, otherwise they would see failures.

earthling-amzn replied:
https://bugs.openjdk.org/browse/JDK-8338695

rkennke commented on test/hotspot/jtreg/gc/stress/jfr/TestStressBigAllocationGCEventsWithShenandoah.java:
```diff
@@ -0,0 +1,50 @@
+  * @test id=generational
+  * @key randomness
+  * @requires vm.hasJFR
+  * @requires vm.gc == "Shenandoah"
```
Same as above.

rkennke replied:
You might want to scan all your new test configs for this pattern.

rkennke replied:
Or actually, you might even want to build and test JDK without Shenandoah to test for other hiccups, like missing INCLUDE_SHENANDOAH_GC, although I don't think there are many places in this PR where this would cause problems.

earthling-amzn replied:
https://bugs.openjdk.org/browse/JDK-8338695

rkennke reviewed (commented):
More reviews. This covers the SA changes and tests.

earthling-amzn commented on test/hotspot/jtreg/gc/shenandoah/oom/TestAllocLargerThanHeap.java:
```diff
@@ -1,76 +0,0 @@
-/*
```
We combined the 'small' and 'large' versions of these tests into a single test case.

rkennke replied:
ok

earthling-amzn commented on test/hotspot/jtreg/gc/shenandoah/oom/TestAllocOutOfMemory.java:
```diff
@@ -0,0 +1,147 @@
+ * @library /test/lib
+ * @run driver TestAllocOutOfMemory small
+ */
+
```
The `main` function in these tests runs default and generational modes:
```java
expectFailure("-Xmx16m",
                          "-XX:+UnlockExperimentalVMOptions",
                          "-XX:+UseShenandoahGC",
                          TestAllocOutOfMemory.class.getName(),
                          "test", size, Integer.toString(heap));

expectFailure("-Xmx16m",
                          "-XX:+UnlockExperimentalVMOptions",
                          "-XX:+UseShenandoahGC", "-XX:ShenandoahGCMode=generational",
                          TestAllocOutOfMemory.class.getName(),
                          "test", size, Integer.toString(heap));
```

rkennke replied:
ok

earthling-amzn commented on test/hotspot/jtreg/gc/shenandoah/oom/TestAllocSmallObj.java:
```diff
@@ -1,80 +0,0 @@
-/*
```
Moved into `TestAllocOutOfMemory.java`.

rkennke replied:
ok

earthling-amzn commented on test/hotspot/jtreg/gc/shenandoah/oom/TestAllocSmallObj.java:
The `small` and `large` tests were rolled into an existing test.

earthling-amzn commented on test/hotspot/jtreg/gc/shenandoah/oom/TestAllocLargeObj.java:
The change was reviewed here: https://github.com/openjdk/shenandoah/pull/233

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp:
```diff
@@ -500,8 +556,7 @@ void ShenandoahConcurrentGC::op_reset() {
   }
-
-  heap->prepare_gc();
+  _generation->prepare_gc();
```
The old generation collection is preceded by a "bootstrap" young collection. We only `prepare_gc` for the old generation during such a bootstrap cycle. Also see, `ShenandoahOldGC::collect` - it's basically just the marking phase, it does not reset the old generation.

rkennke replied:
ok

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp:
```diff
@@ -992,7 +1176,23 @@ void ShenandoahConcurrentGC::op_final_updaterefs() {
+
+  ShenandoahHeap *heap = ShenandoahHeap::heap();
+  heap->set_concurrent_weak_root_in_progress(false);
+  heap->set_evacuation_in_progress(false);
```
This will take a little explaining...

The generational mode allows entire regions to be "promoted in place". This happens during the evacuation phase -_even if there is nothing to evacuate_. So, in the case when the collection set is empty, but we have regions we want to promote in place we turn this flag on to progress through the evacuation phase. However, in this case, we will still skip the update references phase which ordinarily clears the evacuation in progress phase. So, to preserve the expected value of this flag for these short-cut cycles that promote in place, we clear the evacuation in progress flag here in `op_final_roots`.

rkennke replied:
ok, thank you!

rkennke commented on src/hotspot/share/gc/shenandoah/shenandoahGenerationalEvacuationTask.cpp:
```diff
@@ -0,0 +1,288 @@
+
+// When we promote a region in place, we can continue to use the established marking context to guide subsequent remembered
+// set scans of this region's content.  The region will be coalesced and filled prior to the next old-gen marking effort.
+// We identify the entirety of the region as DIRTY to force the next remembered set scan to identify the "interesting poitners"
```
typo: poitners -> pointers

earthling-amzn replied:
Good catch.

rkennke commented on src/hotspot/share/gc/shenandoah/shenandoahGenerationalEvacuationTask.cpp:
```diff
@@ -0,0 +1,288 @@
+  // TODO: Consider not promoting humongous objects that represent primitive arrays.  Leaving a primitive array
+  // (obj->is_typeArray()) in young-gen is harmless because these objects are never relocated and they are not
+  // scanned.  Leaving primitive arrays in young-gen memory allows their memory to be reclaimed more quickly when
+  // it becomes garbage.  Better to not make this change until sizes of young-gen and old-gen are completely
```
So... are sizes of young and old 'completely adaptive' now, and would this optimization make sense?

kdnilsen replied:
I agree.  I believe there are no longer any blockers on this optimization.

earthling-amzn replied:
@earthling-amzn will spin this out into a JBS ticket on his PR to clean up TODOs.

kdnilsen replied:
Tracking this under https://bugs.openjdk.org/browse/JDK-8339612

rkennke commented on src/hotspot/share/gc/shenandoah/shenandoahGenerationalFullGC.cpp:
```diff
@@ -0,0 +1,388 @@
+    if (_old_compact_point + obj_size > _old_to_region->end()) {
+      ShenandoahHeapRegion* new_to_region;
+
+      log_debug(gc)("Worker %u finishing old region " SIZE_FORMAT ", compact_point: " PTR_FORMAT ", obj_size: " SIZE_FORMAT
```
This evaluates whether or not logging is set to 'debug' for every single live object, right? Might want to change this into log_develop_debug() so that release builds are not impacted by that (and would not log that). There are more instances of this problem in this method.

earthling-amzn replied:
@earthling-amzn will do this.

earthling-amzn replied:
On a closer look, this only happens when a region is finished so it is not on the order of the number of oops. However, after reviewing other uses of `log_debug` I did find a few that would be better as `log_develop_debug`. Also, there are quite some `log_trace` statements in shReferenceProcessor that probably should be `log_develop_trace`, but I'd do that in a separate, unrelated PR.

https://bugs.openjdk.org/browse/JDK-8338779

rkennke commented on src/hotspot/share/gc/shenandoah/shenandoahOldGC.hpp:
```diff
@@ -0,0 +1,48 @@
+class ShenandoahOldGC : public ShenandoahConcurrentGC {
+ public:
+  ShenandoahOldGC(ShenandoahOldGeneration* generation, ShenandoahSharedFlag& allow_preemption);
+  bool collect(GCCause::Cause cause);
```
You might want to mark this as override.

earthling-amzn replied:
Yes.

rkennke commented on src/hotspot/share/gc/shenandoah/shenandoahOldGC.hpp:
```diff
@@ -0,0 +1,48 @@
+  bool collect(GCCause::Cause cause);
+
+ protected:
+  virtual void op_final_mark();
```
You might want to mark this as override, and remove virtual.

earthling-amzn replied:
Will do.

rkennke reviewed (commented):
More reviews (before GitHub goes down again ;-) ), this time most of new files. A few large ones are left, I'll look at those tomorrow.

ysramakrishna commented on src/hotspot/share/gc/shenandoah/shenandoahAgeCensus.cpp:
```diff
@@ -0,0 +1,383 @@
+NO_CENSUS_NOISE(void ShenandoahAgeCensus::add(uint obj_age, uint region_age, size_t size, uint worker_id) {)
+  if (obj_age <= markWord::max_age) {
+    assert(obj_age < MAX_COHORTS && region_age < MAX_COHORTS, "Should have been tenured");
+#ifdef SHENANDOAH_CENSUS_NOISE
```
The idea was to have this be independently controllable:
```
#ifndef PRODUCT
// Enable noise instrumentation
#define SHENANDOAH_CENSUS_NOISE 1
#endif  // PRODUCT

#ifdef SHENANDOAH_CENSUS_NOISE

#define CENSUS_NOISE(x) x
#define NO_CENSUS_NOISE(x)
...
#else  // SHENANDOAH_CENSUS_NOISE
#define CENSUS_NOISE(x)
#define NO_CENSUS_NOISE(x) x
#endif // SHENANDOAH_CENSUS_NOISE
```

The instrumentation is present only in non-product (non-release) builds. It's thus tested ony in debug/fastdebug mode, but can be turned on, if needed in release mode when doing specific debugging, by means of editing the above.

Since both paths are tested (one in fastdebug mode and the other in release mode), all code paths are always tested and won't bit-rot.

rkennke replied:
ok, then

rkennke commented on src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.hpp:
```diff
@@ -0,0 +1,1062 @@
+  inline bool has_next() const;
+};
+
+typedef ShenandoahScanRemembered<ShenandoahDirectCardMarkRememberedSet> RememberedScanner;
```
Do I see this correctly that there is only one implementation of the template parameter that is passed around all over this and related source files? If yes, then we could just as well clean this up, remove all the templating and move much/all of the code out of .inline.hpp and into .cpp file instead, and avoid making it inline, which might be counter-productive.

ysramakrishna replied:
Tracking this work in ticket https://bugs.openjdk.org/browse/JDK-8338479

earthling-amzn replied:
https://github.com/openjdk/shenandoah/pull/472

rkennke reviewed (changes requested):
Final part of my first-pass review.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahAsserts.cpp:
```diff
@@ -425,3 +428,20 @@ void ShenandoahAsserts::assert_heaplocked_or_safepoint(const char* file, int lin
 }
+
+// Unlike assert_heaplocked_or_safepoint(), this does not require current thread in safepoint to be a VM thread
+// TODO: This should be more aptly named. Nothing in this method checks we are actually in Full GC.
```
Code is no longer using the version of the assert that checks that the current thread is the VMThread.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.hpp:
```diff
@@ -27,6 +28,7 @@
 
 #include "gc/shared/barrierSet.hpp"
 #include "gc/shared/bufferNode.hpp"
+#include "gc/shenandoah/shenandoahCardTable.hpp"
```
This is done.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.cpp:
```diff
@@ -52,6 +54,10 @@ ShenandoahBarrierSet::ShenandoahBarrierSet(ShenandoahHeap* heap) :
   _satb_mark_queue_buffer_allocator("SATB Buffer Allocator", ShenandoahSATBBufferSize),
   _satb_mark_queue_set(&_satb_mark_queue_buffer_allocator)
```
Done.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp:
```diff
@@ -860,6 +1185,7 @@ HeapWord* ShenandoahFreeSet::allocate_contiguous(ShenandoahAllocRequest& req) {
   }
 
   size_t remainder = words_size & ShenandoahHeapRegion::region_size_words_mask();
+  bool is_generational = _heap->mode()->is_generational();
```
I'll remove it.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahFullGC.cpp:
```diff
@@ -168,10 +179,10 @@ void ShenandoahFullGC::do_it(GCCause::Cause gc_cause) {
     if (heap->is_concurrent_mark_in_progress()) {
-      ShenandoahConcurrentGC::cancel();
-      heap->set_concurrent_mark_in_progress(false);
+      // TODO: Send cancel_concurrent_mark upstream? Does it really not have it already?
```
Just a note to myself. I've removed this.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahFullGC.cpp:
```diff
@@ -181,24 +192,29 @@ void ShenandoahFullGC::do_it(GCCause::Cause gc_cause) {
   }
 
   if (UseTLAB) {
+    // TODO: Do we need to explicitly retire PLABs?
```
Updated this comment (PLABs are retired alongside GCLABs).

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahFullGC.cpp:
```diff
@@ -730,6 +773,8 @@ void ShenandoahFullGC::phase2_calculate_target_addresses(ShenandoahHeapRegionSet
 
     distribute_slices(worker_slices);
 
+    // TODO: This is ResourceMark is missing upstream.
```
It was not needed. It's been removed.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:
```diff
@@ -216,6 +225,28 @@ jint ShenandoahHeap::initialize() {
                               "Cannot commit heap memory");
   }
 
+  BarrierSet::set_barrier_set(new ShenandoahBarrierSet(this, _heap_region));
```
It is because `ShenandoahBarrierSet` now hosts `ShenandoahCardTable`. `ShenandoahCardTable` needs to know the size and location of the heap. Nothing will use the barrier set at this point.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:
```diff
@@ -354,11 +385,13 @@ jint ShenandoahHeap::initialize() {
   {
     ShenandoahHeapLocker locker(lock());
 
+
```
Agree

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:
```diff
@@ -683,33 +684,84 @@ void ShenandoahHeap::decrease_committed(size_t bytes) {
+void ShenandoahHeap::increase_humongous_waste(ShenandoahGeneration* generation, size_t bytes) {
+  generation->increase_humongous_waste(bytes);
+  if (!generation->is_global()) {
+    global_generation()->increase_humongous_waste(bytes);
```
That could be done, but we'd need to make these methods `virtual` and would have to evaluate the performance impact.

rkennke replied:
Is that so hot, though? The only caller seems to be ShHeap::allocate_memory() which has a lot of stuff going on that is surely more complex than some virtual calls. I wouldn't be *too* worried about virtual calls, unless you're in a really really tight loop.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:
```diff
@@ -1015,7 +1069,43 @@ HeapWord* ShenandoahHeap::allocate_memory_under_lock(ShenandoahAllocRequest& req
-  return _free_set->allocate(req, in_new_region);
+
+  // Make sure the old generation has room for either evacuations or promotions before trying to allocate.
+  if (req.is_old() && !old_generation()->can_allocate(req)) {
```
Yes, but we'd then need to make this method `virtual` and we've been reluctant to do such things on a hot path, like allocation. Perhaps this reluctance is misguided?

rkennke replied:
Virtual calls are quite fast, they are basically indirect jumps (as opposed to direct jumps for non-virtual calls). This allocation path is not all that hot, either. It's the 'slow' path that serves out TLABs and larger shared allocs, and goes under a lock and has a bunch of other complexity which surely trumps a virtual call.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:
```diff
@@ -1152,11 +1257,21 @@ oop ShenandoahHeap::evacuate_object(oop p, Thread* thread) {
+      if ((copy == nullptr) && (size < ShenandoahThreadLocalData::gclab_size(thread))) {
+        // GCLAB allocation failed because we are bumping up against the limit on young evacuation reserve.  Try resetting
+        // the desired GCLAB size and retry GCLAB allocation to avoid cascading of shared memory allocations.
+        // TODO: is this right? using PLAB::min_size() here for gc lab size?
```
Yes, this `TODO` has been removed.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahHeap.hpp:
```diff
@@ -348,8 +418,11 @@ class ShenandoahHeap : public CollectedHeap, public ShenandoahSpaceInfo {
+  bool is_prepare_for_old_mark_in_progress() const;
 
 private:
+  void manage_satb_barrier(bool active);
```
It guards the action to propagate activating/deactivating the barrier to all the threads (this gets a little tricky with concurrent marking of old/young). Could call it `maybe_toggle_satb_barrier`?

rkennke replied:
yeah that'd be better.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.hpp:
```diff
@@ -355,7 +384,31 @@ class ShenandoahHeapRegion {
+  // Coalesce contiguous spans of garbage objects by filling header and reregistering start locations with remembered set.
+  // This is used by old-gen GC following concurrent marking to make old-gen HeapRegions parsable.  Return true iff
+  // region is completely coalesced and filled.  Returns false if cancelled before task is complete.
+  bool oop_coalesce_and_fill(bool cancellable);
```
Yes, the classes may be unloaded, but it's more likely these dead objects will hold pointers into evacuated memory. Being dead, these objects will not have such references updated.

kdnilsen replied:
There is also a "performance" benefit in that subsequent remembered set scanning does not have to individually examine every previously allocated dead object.  Rather, N continguous dead objects are skipped over en masse.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahHeapRegionCounters.cpp:
```diff
@@ -21,15 +22,18 @@
  * questions.
  *
  */
-
```
Okay

earthling-amzn commented on test/hotspot/jtreg/compiler/c2/irTests/TestIfMinMax.java:
```diff
@@ -30,7 +30,7 @@
 /*
  * @test
- * @bug 8324655 8329797
+ * @bug 8324655 8329797 8331090
```
This should be fixed. This file matches the [target branch now](https://github.com/openjdk/jdk/blob/master/test/hotspot/jtreg/compiler/c2/irTests/TestIfMinMax.java#L33).

earthling-amzn commented on src/hotspot/share/gc/shenandoah/heuristics/shenandoahHeapStats.hpp:
```diff
@@ -0,0 +1,41 @@
+ *
+ */
+
+#ifndef SHARE_GC_SHENANDOAH_HEURISTICS_SHENANDOAHHEAPCHARACTERISTICS_HPP
```
This file shouldn't be there. Deleted.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahCardTable.hpp:
```diff
@@ -0,0 +1,90 @@
+class ShenandoahCardTable: public CardTable {
+  friend class VMStructs;
+
+protected:
```
Agree

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahCardTable.hpp:
```diff
@@ -0,0 +1,90 @@
+public:
+  ShenandoahCardTable(MemRegion whole_heap) : CardTable(whole_heap) { }
+
+  virtual void initialize();
```
Agree

ysramakrishna commented on src/hotspot/share/gc/shenandoah/shenandoahCollectionSetPreselector.hpp:
```diff
@@ -0,0 +1,48 @@
+  }
+
+  ~ShenandoahCollectionSetPreselector() {
+    _cset->abandon_preselected();
```
No, see discussion at https://github.com/openjdk/jdk/pull/20395/files#r1699117626. The array being resource allocated, is freed when the enclosing ResourceMark is exited. There is therefore no leak.

This should go away when https://bugs.openjdk.org/browse/JDK-8338472 is done.

Resolving this thread.

ysramakrishna commented on src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp:
```diff
@@ -112,9 +116,19 @@ bool ShenandoahConcurrentGC::collect(GCCause::Cause cause) {
+    TASKQUEUE_STATS_ONLY(_mark.task_queues()->reset_taskqueue_stats());
+
+    // Concurrent remembered set scanning
+    entry_scan_remembered_set();
```
Tracked in https://bugs.openjdk.org/browse/JDK-8338475.
Resolving comment thread.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahGenerationalHeap.cpp:
```diff
@@ -0,0 +1,1127 @@
+
+void ShenandoahGenerationalHeap::initialize_controller() {
+  auto control_thread = new ShenandoahGenerationalControlThread();
+  _control_thread = control_thread;
```
`ShenandoahHeap::_control_thread` has the type `ShenandoahController`, but the constructor for `ShenandoahRegularThread` requires `ShenandoahGenerationalControlThread`. Using a local variable here with type `ShenandoahGenerationalControlThread` avoids having to downcast.

rkennke replied:
`ShenandoahHeap::_control_thread` has the type `ShenandoahController*`, and `control_thread` will get the type `ShenandoahGenerationalControlThread*`. You assign a `ShenandoahGenerationalControlThread*` to `ShenandoahHeap::_control_thread`, no matter which way you write it, and that assignment is 100% legal.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahGenerationalHeap.cpp:
```diff
@@ -0,0 +1,1127 @@
+          break;
+        }
+        case OLD_GENERATION: {
+          assert(mode()->is_generational(), "OLD Generation only exists in generational mode");
```
Yes, this assert is a vestige from an earlier form of this code.

earthling-amzn replied:
https://github.com/openjdk/shenandoah/pull/474

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahGenerationalHeap.cpp:
```diff
@@ -0,0 +1,1127 @@
+  }
+}
+
+namespace ShenandoahCompositeRegionClosure {
```
It is perhaps not strictly _required_ to use a namespace here. You would prefer no containing namespace? or a class with a static function and an inner class?

rkennke replied:
I guess it's ok.

kdnilsen commented on src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:
```diff
@@ -370,12 +403,18 @@ jint ShenandoahHeap::initialize() {
+    // We are initializing free set.  We ignore cset region tallies.
+    size_t first_old, last_old, num_old;
+    _free_set->prepare_to_rebuild(young_cset_regions, old_cset_regions, first_old, last_old, num_old);
+    _free_set->finish_rebuild(young_cset_regions, old_cset_regions, num_old);
```
In single-generation mode, there is no need to separate these.  But in generational mode, we sometimes desire to resize the generations before rebuilding, and the parameters that constrain resizing of the generations are gathered by prepare_to_rebuild().

kdnilsen commented on src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp:
```diff
@@ -354,11 +385,13 @@ jint ShenandoahHeap::initialize() {
   }
 
   _regions = NEW_C_HEAP_ARRAY(ShenandoahHeapRegion*, _num_regions, mtGC);
+  _affiliations = NEW_C_HEAP_ARRAY(uint8_t, _num_regions, mtGC);
```
This was measured to offer a performance benefit on several tests.  Fetching affiliation through the ShHeapRegion requires two levels of indirection (consult array to find the heap region, consult the heap region to find the field), and less cache locality.  There are many loops that iterate over regions and treat regions of different affiliation (young, old, free) differently.

kdnilsen commented on src/hotspot/share/gc/shenandoah/heuristics/shenandoahHeuristics.cpp:
```diff
@@ -24,32 +25,33 @@
+// sort by decreasing garbage (so most garbage comes first)
 int ShenandoahHeuristics::compare_by_garbage(RegionData a, RegionData b) {
-  if (a._garbage > b._garbage)
+  if (a._u._garbage > b._u._garbage)
```
I will create accessors to hide this union.

ysramakrishna replied:
@kdnilsen : https://github.com/openjdk/shenandoah/pull/475

kdnilsen commented on src/hotspot/share/gc/shenandoah/shenandoahCollectionSet.hpp:
```diff
@@ -33,6 +34,14 @@
   friend class ShenandoahHeap;
+  friend class ShenandoahCollectionSetPreselector;
+
+  void establish_preselected(bool *preselected) {
```
I will also make this change.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahGenerationalHeap.cpp:
```diff
@@ -0,0 +1,1127 @@
+  }
+};
+
+ShenandoahGenerationalHeap* ShenandoahGenerationalHeap::heap() {
```
https://github.com/openjdk/shenandoah/pull/474

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahGenerationalHeap.cpp:
```diff
@@ -0,0 +1,1127 @@
+  return cast(heap);
+}
+
+ShenandoahGenerationalHeap* ShenandoahGenerationalHeap::cast(CollectedHeap* heap) {
```
https://github.com/openjdk/shenandoah/pull/474

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahGenerationalHeap.cpp:
```diff
@@ -0,0 +1,1127 @@
+private:
+  template<class T>
+  void do_work(uint worker_id) {
+    T cl;
```
https://github.com/openjdk/shenandoah/pull/474

shipilev commented on src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.inline.hpp:
```diff
@@ -323,16 +364,23 @@ template <typename T>
+  bs->arraycopy_barrier(src, dst, length);
+  bool result = Raw::oop_arraycopy_in_heap(src_obj, src_offset_in_bytes, src_raw, dst_obj, dst_offset_in_bytes, dst_raw, length);
+  if (ShenandoahCardBarrier) {
+    bs->write_ref_array((HeapWord*) dst, length);
```
Um. So we are here, where `T` might mean either `oop` or `narrowOop`. Yet, we are calling `write_ref_array` with `HeapWord*` and `length`, which is here a _count_ of either `oop` or `narrowOop`-s. So for `narrowOop`, which is smaller than `HeapWord`, we call `write_ref_array` with _double_ the length? 

Note how `ShenandoahBarrierSet::arraycopy_barrier` carries `T* src, size_t count` all the way down to `ShenandoahBarrierSet::arraycopy_work` to avoid this problem. I think `write_ref_array` should really carry overloads for `write_ref_array(oop* src, size_t count)` and `write_ref_array(narrowOop* src, size_t count)`.

earthling-amzn replied:
@kdnilsen will take a closer look.

kdnilsen replied:
Tracking this in https://bugs.openjdk.org/browse/JDK-8339615

kdnilsen replied:
Inside of write_ref_array(), we calculate the end of the array using:
```
  HeapWord* end = (HeapWord*)((char*) start + (count * heapOopSize));
```
Then we mark between array start and array end as dirty.  Though this code as written is a "bit" clumsy, I believe it is correct.

kdnilsen replied:
Am willing to try cleaning up the code, but am concerned this introduces risk of introducing regressions.  I see that the same pattern appears in the generic ModRefBarrierSet::write_ref_array() and in ModRefBarrierSet::AccessBarrier::oop_arraycopy_in_heap().

shipilev replied:
All right.

shipilev commented on test/hotspot/jtreg/gc/shenandoah/TestPeriodicGC.java:
```diff
@@ -45,14 +46,39 @@ public static void testWith(String msg, boolean periodic, String... args) throws
+        OutputAnalyzer output = new OutputAnalyzer(pb.start());
+        output.shouldHaveExitValue(0);
+        if (periodic) {
+            if (!output.getOutput().contains("Trigger (YOUNG): Time since last GC")) {
+                throw new AssertionError("Generational mode: Should have periodic young GC in logs");
+            }
```
Here, and in other tests, use `output.shouldContain(...)`

earthling-amzn replied:
https://bugs.openjdk.org/browse/JDK-8338695

shipilev commented on test/hotspot/jtreg/gc/shenandoah/TestLargeObjectAlignment.java:
```diff
@@ -23,7 +24,7 @@
 
 /*
- * @test
+ * @test default
```
`id=default`

earthling-amzn replied:
https://bugs.openjdk.org/browse/JDK-8338695

shipilev commented on test/hotspot/jtreg/gc/shenandoah/TestHumongousThreshold.java:
```diff
@@ -108,6 +109,85 @@
  *                   TestHumongousThreshold
  */
 
+/*
```
This test should be gone after [JDK-8338444](https://bugs.openjdk.org/browse/JDK-8338444).

earthling-amzn replied:
For historical reasons, `shenandoah:master` usually only merges the weekly tags from `openjdk:master`.

shipilev reviewed (commented):
Some comments for tests as well:

kdnilsen commented on src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp:
```diff
@@ -671,109 +840,249 @@ HeapWord* ShenandoahFreeSet::allocate_single(ShenandoahAllocRequest& req, bool&
+// at the start of the free space.
+//
+// This is merely a helper method to use for the purpose of such a calculation.
+size_t ShenandoahFreeSet::get_usable_free_words(size_t free_bytes) const {
```
Do we want to universally change all byte accounting to words?  before upstream merge?  If so, I'll work on this.

kdnilsen commented on src/hotspot/share/gc/shenandoah/shenandoahFullGC.cpp:
```diff
@@ -255,6 +271,7 @@ void ShenandoahFullGC::do_it(GCCause::Cause gc_cause) {
     heap->verifier()->verify_after_fullgc();
   }
 
+  // Humongous regions are promoted on demand and are accounted for by normal Full GC mechanisms.
```
Agree this seems out of place and of no value.  Will remove.

rkennke replied:
ok

earthling-amzn commented on test/hotspot/jtreg/gc/shenandoah/generational/TestCLIModeGenerational.java:
```diff
@@ -0,0 +1,54 @@
+ *      gc.shenandoah.generational.TestCLIModeGenerational
+ */
+
+public class TestCLIModeGenerational {
+
+  private static WhiteBox wb = WhiteBox.getWhiteBox();
+
+  public static void main(String args[]) throws Exception {
+    Boolean using_shenandoah = wb.getBooleanVMFlag("UseShenandoahGC");
+    String gc_mode = wb.getStringVMFlag("ShenandoahGCMode");
+    if (!using_shenandoah || !gc_mode.equals("generational"))
+      throw new IllegalStateException("Command-line options not honored!");
+  }
+}
```
We will roll this test into another test that asserts that the expected flags are enabled and disabled for different modes.

earthling-amzn commented on test/hotspot/jtreg/gc/shenandoah/generational/TestOldGrowthTriggers.java:
```diff
@@ -0,0 +1,115 @@
+    ProcessBuilder pb = ProcessTools.createLimitedTestJavaProcessBuilder(cmds);
+    OutputAnalyzer output = new OutputAnalyzer(pb.start());
+    output.shouldHaveExitValue(0);
+    if (!output.getOutput().contains("Trigger (OLD): Old has overgrown")) {
+      throw new AssertionError("Generational mode: Should experience OLD triggers due to growth");
+    }
```
@kdnilsen will fix up this test case.

earthling-amzn commented on test/hotspot/jtreg/gc/shenandoah/jni/TestPinnedGarbage.java:
```diff
@@ -61,6 +62,22 @@
+ * @requires vm.gc.Shenandoah
+ * @library /test/lib
+ *
+ * @run main/othervm/native -XX:+UnlockDiagnosticVMOptions -XX:+UnlockExperimentalVMOptions -Xmx128m
+ *      -XX:+UseShenandoahGC -XX:ShenandoahGCMode=generational
+ *      TestPinnedGarbage
+ *
+ * @run main/othervm/native -XX:+UnlockDiagnosticVMOptions -XX:+UnlockExperimentalVMOptions -Xmx128m
+ *      -XX:+UseShenandoahGC -XX:ShenandoahGCMode=generational
+ *      -XX:+ShenandoahVerify
+ *      TestPinnedGarbage
+ */
```
@ysramakrishna will go over the tests and make sure verifier is enabled in the first round.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.inline.hpp:
```diff
@@ -274,15 +297,24 @@ template <typename T>
   ShenandoahBarrierSet* bs = ShenandoahBarrierSet::barrier_set();
-  return bs->oop_cmpxchg(decorators, addr, compare_value, new_value);
+  oop result = bs->oop_cmpxchg(decorators, addr, compare_value, new_value);
+  if (ShenandoahCardBarrier) {
```
@earthling-amzn will take a closer look here.

kdnilsen replied:
Tracking this in https://bugs.openjdk.org/browse/JDK-8339613

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.cpp:
```diff
@@ -124,6 +132,16 @@ void ShenandoahBarrierSet::on_thread_detach(Thread *thread) {
 
+    if (ShenandoahCardBarrier) {
+      PLAB* plab = ShenandoahThreadLocalData::plab(thread);
+      // retire_plab may register the remnant filler object with the remembered set scanner without a lock.
+      // This is safe because it is assured that each PLAB is a whole-number multiple of card-mark memory size and each
+      // PLAB is aligned with the start of each card's memory range.
```
@kdnilsen will find the right home for this comment.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahAsserts.hpp:
```diff
@@ -163,6 +164,27 @@ class ShenandoahAsserts {
 #define shenandoah_assert_heaplocked_or_safepoint() \
                     ShenandoahAsserts::assert_heaplocked_or_safepoint(__FILE__, __LINE__)
+
+#define shenandoah_assert_control_or_vm_thread() \
```
@ysramakrishna will fix this.

earthling-amzn replied:
https://github.com/openjdk/shenandoah/pull/478

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahArguments.cpp:
```diff
@@ -68,6 +71,13 @@ void ShenandoahArguments::initialize() {
+  // We use this as the time period for tracking minimum mutator utilization (MMU).
+  // In generational mode, the MMU is used as a signal to adjust the size of the
+  // young generation.
+  if (FLAG_IS_DEFAULT(GCPauseIntervalMillis)) {
+    FLAG_SET_DEFAULT(GCPauseIntervalMillis, 5000);
+  }
```
@earthling-amzn will fix this. This is somewhat vestigial.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahAllocRequest.hpp:
```diff
@@ -47,64 +50,95 @@ class ShenandoahAllocRequest : StackObj {
   }
 
 private:
+  // When ShenandoahElasticTLAB is enabled, the request cannot be made smaller than _min_size.
```
Yes, @earthling-amzn will fix this comment.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/heuristics/shenandoahOldHeuristics.cpp:
```diff
@@ -0,0 +1,713 @@
+
+// sort by increasing live (so least live comes first)
+int ShenandoahOldHeuristics::compare_by_live(RegionData a, RegionData b) {
+  if (a._u._live_data < b._u._live_data)
+    return -1;
+  else if (a._u._live_data > b._u._live_data)
+    return 1;
+  else return 0;
```
@kdnilsen has added braces in the most recent version of this code (pending merge).

earthling-amzn commented on src/hotspot/share/gc/shenandoah/heuristics/shenandoahGlobalHeuristics.cpp:
```diff
@@ -0,0 +1,176 @@
+  for (size_t idx = 0; idx < size; idx++) {
+    ShenandoahHeapRegion* r = data[idx]._region;
+    if (cset->is_preselected(r->index())) {
+      fatal("There should be no preselected regions during GLOBAL GC");
```
@kdnilsen could change this to an assert.

ysramakrishna replied:
@ysramakrishna is changing this to an assert as I am touching things in the vicinity of this. I will however have a question about this in the upcoming PR, associated with https://bugs.openjdk.org/browse/JDK-8338780. Stay tuned.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/heuristics/shenandoahGlobalHeuristics.cpp:
```diff
@@ -0,0 +1,176 @@
+void ShenandoahGlobalHeuristics::choose_collection_set_from_regiondata(ShenandoahCollectionSet* cset,
+                                                                       RegionData* data, size_t size,
+                                                                       size_t actual_free) {
+  // The logic for cset selection in adaptive is as follows:
```
@ysramakrishna - will clean up this comment or move it out of line.

ysramakrishna replied:
@ysramakrishna : https://bugs.openjdk.org/browse/JDK-8338780

earthling-amzn commented on src/hotspot/share/gc/shenandoah/heuristics/shenandoahAdaptiveHeuristics.hpp:
```diff
@@ -138,6 +139,8 @@ class ShenandoahAdaptiveHeuristics : public ShenandoahHeuristics {
   // establishes what is 'normal' for the application and is used as a
   // source of feedback to adjust trigger parameters.
   TruncatedSeq _available;
+
+  size_t min_free_threshold();
```
@ysramakrishna will add a small comment here.

ysramakrishna replied:
@ysramakrishna : https://bugs.openjdk.org/browse/JDK-8338780

earthling-amzn commented on src/hotspot/share/gc/shenandoah/heuristics/shenandoahAdaptiveHeuristics.cpp:
```diff
@@ -196,41 +206,68 @@ static double saturate(double value, double min, double max) {
     }
   }
-
+  //  Rationale:
```
Not specific to GenShen, @ysramakrishna will move this out of line.

ysramakrishna replied:
@shipilev : I can certainly move this out of line if it detracts from reading the code. There is no change in the code itself, just the comment really and minor changes in the logging messages. Is that what you'd like done here, or did you have something else in mind?

ysramakrishna replied:
@ysramakrishna : https://bugs.openjdk.org/browse/JDK-8338780

earthling-amzn commented on src/hotspot/share/gc/shenandoah/heuristics/shenandoahAdaptiveHeuristics.cpp:
```diff
@@ -130,17 +137,20 @@ void ShenandoahAdaptiveHeuristics::record_success_concurrent() {
   double z_score = 0.0;
-  if (_available.sd() > 0) {
-    z_score = (available - _available.avg()) / _available.sd();
+  double available_sd = _available.sd();
```
@earthling-amzn will try to upstream these general changes separately.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/heuristics/shenandoahAdaptiveHeuristics.cpp:
```diff
@@ -196,41 +206,68 @@ static double saturate(double value, double min, double max) {
-  // Make sure the code below treats available without the soft tail.
-  size_t soft_tail = max_capacity - capacity;
-  available = (available > soft_tail) ? (available - soft_tail) : 0;
+  log_debug(gc)("should_start_gc (%s)? available: " SIZE_FORMAT ", soft_max_capacity: " SIZE_FORMAT
```
We haven't looked for this message outside of honest debugging efforts.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/heuristics/shenandoahAdaptiveHeuristics.cpp:
```diff
@@ -86,11 +93,11 @@ void ShenandoahAdaptiveHeuristics::choose_collection_set_from_regiondata(Shenand
   size_t max_cset    = (size_t)((1.0 * capacity / 100 * ShenandoahEvacReserve) / ShenandoahEvacWaste);
-  size_t free_target = (capacity / 100 * ShenandoahMinFreeThreshold) + max_cset;
-  size_t min_garbage = (free_target > actual_free ? (free_target - actual_free) : 0);
+  size_t free_target = (capacity * ShenandoahMinFreeThreshold) / 100 + max_cset;
+  size_t min_garbage = (free_target > actual_free) ? (free_target - actual_free) : 0;
```
@kdnilsen - will upstream this separately.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/heuristics/shenandoahAdaptiveHeuristics.cpp:
```diff
@@ -86,11 +93,11 @@ void ShenandoahAdaptiveHeuristics::choose_collection_set_from_regiondata(Shenand
 
   log_info(gc, ergo)("Adaptive CSet Selection. Target Free: " SIZE_FORMAT "%s, Actual Free: "
-                     SIZE_FORMAT "%s, Max CSet: " SIZE_FORMAT "%s, Min Garbage: " SIZE_FORMAT "%s",
+                     SIZE_FORMAT "%s, Max Evacuation: " SIZE_FORMAT "%s, Min Garbage: " SIZE_FORMAT "%s",
```
This was meant to be less confusing to the reader of the log. Evacuation is generally a more well understood concept than "CSet".

earthling-amzn commented on src/hotspot/share/gc/shenandoah/c2/shenandoahBarrierSetC2.cpp:
```diff
@@ -847,9 +959,25 @@ void ShenandoahBarrierSetC2::unregister_potential_barrier_node(Node* node) const
-    shenandoah_eliminate_wb_pre(n, &macro->igvn());
+void ShenandoahBarrierSetC2::eliminate_gc_barrier(PhaseMacroExpand* macro, Node* node) const {
+  if (is_shenandoah_wb_pre_call(node)) {
+    shenandoah_eliminate_wb_pre(node, &macro->igvn());
```
Good question! We will take a closer look.

ysramakrishna replied:
Will be considered in https://bugs.openjdk.org/browse/JDK-8339521.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/c2/shenandoahBarrierSetC2.cpp:
```diff
@@ -485,8 +571,22 @@ Node* ShenandoahBarrierSetC2::store_at_resolved(C2Access& access, C2AccessValue&
+    return result;
+  } else {
+    assert(access.is_opt_access(), "only for optimization passes");
+    assert(((decorators & C2_TIGHTLY_COUPLED_ALLOC) != 0 || !ShenandoahSATBBarrier) && (decorators & C2_ARRAY_COPY) != 0, "unexpected caller of this code");
```
@ysramakrishna will investigate.

ysramakrishna replied:
Will be considered in https://bugs.openjdk.org/browse/JDK-8339521.

earthling-amzn commented on src/hotspot/cpu/x86/gc/shenandoah/shenandoahBarrierSetAssembler_x86.cpp:
```diff
@@ -120,6 +122,29 @@ void ShenandoahBarrierSetAssembler::arraycopy_prologue(MacroAssembler* masm, Dec
+    if (ShenandoahCardBarrier) {
+      bool checkcast = (decorators & ARRAYCOPY_CHECKCAST) != 0;
+      bool disjoint = (decorators & ARRAYCOPY_DISJOINT) != 0;
+      bool obj_int = type == T_OBJECT LP64_ONLY(&& UseCompressedOops);
```
@earthling-amzn will check how other collectors handle this scenario.

earthling-amzn commented on src/hotspot/cpu/x86/gc/shenandoah/c1/shenandoahBarrierSetC1_x86.cpp:
```diff
@@ -84,9 +85,14 @@ LIR_Opr ShenandoahBarrierSetC1::atomic_cmpxchg_at_resolved(LIRAccess& access, LI
       return result;
     }
   }
+
```
@earthling-amzn will remove.

earthling-amzn commented on src/hotspot/cpu/aarch64/gc/shenandoah/shenandoahBarrierSetAssembler_aarch64.cpp:
```diff
@@ -362,6 +371,26 @@ void ShenandoahBarrierSetAssembler::load_at(MacroAssembler* masm, DecoratorSet d
 }
 
+void ShenandoahBarrierSetAssembler::store_check(MacroAssembler* masm, Register obj) {
+  assert(ShenandoahCardBarrier, "Did you mean to enable ShenandoahCardBarrier?");
```
@ysramakrishna will fix this.

earthling-amzn replied:
https://github.com/openjdk/shenandoah/pull/478

earthling-amzn commented on src/hotspot/cpu/aarch64/gc/shenandoah/shenandoahBarrierSetAssembler_aarch64.cpp:
```diff
@@ -581,6 +613,35 @@ void ShenandoahBarrierSetAssembler::cmpxchg_oop(MacroAssembler* masm,
+  __ load_byte_map_base(scratch);
+  __ add(start, start, scratch);
+  __ bind(L_loop);
+  __ strb(zr, Address(start, count));
```
@ysramakrishna will link this up to the existing ticket to remove `UseCondCardMark` and add the code for conditional card marking.

ysramakrishna replied:
Will be considered in https://bugs.openjdk.org/browse/JDK-8339521.

earthling-amzn commented on src/hotspot/cpu/aarch64/gc/shenandoah/shenandoahBarrierSetAssembler_aarch64.cpp:
```diff
@@ -397,6 +426,9 @@ void ShenandoahBarrierSetAssembler::store_at(MacroAssembler* masm, DecoratorSet
     }
     BarrierSetAssembler::store_at(masm, decorators, type, Address(tmp3, 0), val, noreg, noreg, noreg);
+    if (ShenandoahCardBarrier) {
+      store_check(masm, r3);
```
@earthling-amzn will look into this.

earthling-amzn replied:
Looks like we missed this when https://bugs.openjdk.org/browse/JDK-8292868 was integrated.

earthling-amzn commented on src/hotspot/cpu/aarch64/gc/shenandoah/c1/shenandoahBarrierSetC1_aarch64.cpp:
```diff
@@ -86,9 +87,14 @@ LIR_Opr ShenandoahBarrierSetC1::atomic_cmpxchg_at_resolved(LIRAccess& access, LI
       return result;
     }
   }
+
```
@kdnilsen will add to his formatting PR.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahEvacTracker.hpp:
```diff
@@ -0,0 +1,85 @@
+#define SHARE_GC_SHENANDOAH_SHENANDOAHEVACTRACKER_HPP
+
+#include "gc/shared/ageTable.hpp"
+#include "utilities/ostream.hpp"
```
@earthling-amzn will take a look.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/heuristics/shenandoahOldHeuristics.cpp:
```diff
@@ -0,0 +1,722 @@
+    //   else if old_gen spans more than 87.5% and density < 62.5%, trigger old-defrag
+    //   else if old_gen spans more than 75% and density < 50%, trigger old-defrag
+    //   else if old_gen spans more than 62.5% and density < 37.5%, trigger old-defrag
+    //   else if old_gen spans more than 50% and density < 25%, trigger old-defrag
```
@kdnilsen will find a way to make this cleaner.

kdnilsen replied:
Tracking this under issue https://bugs.openjdk.org/browse/JDK-8339611

earthling-amzn commented on test/hotspot/jtreg/gc/shenandoah/TestSieveObjects.java:
```diff
@@ -121,7 +144,7 @@
  * @library /test/lib
  *
- * @run main/othervm/timeout=240 -Xmx1g -Xms1g -XX:+UnlockDiagnosticVMOptions -XX:+UnlockExperimentalVMOptions
+ * @run main/othervm/timeout=300 -Xmx1g -Xms1g -XX:+UnlockDiagnosticVMOptions -XX:+UnlockExperimentalVMOptions
```
It's probably vestigial, @earthling-amzn will try reverting this.

earthling-amzn commented on test/hotspot/jtreg/gc/shenandoah/TestRetainObjects.java:
```diff
@@ -108,7 +134,7 @@
  * @requires vm.gc.Shenandoah
  *
- * @run main/othervm -Xmx1g -Xms1g -XX:+UnlockDiagnosticVMOptions -XX:+UnlockExperimentalVMOptions
+ * @run main/othervm/timeout=300 -Xmx1g -Xms1g -XX:+UnlockDiagnosticVMOptions -XX:+UnlockExperimentalVMOptions
```
@earthling-amzn  should be able to revert this.

earthling-amzn commented on test/hotspot/jtreg/gc/shenandoah/TestAllocIntArrays.java:
```diff
@@ -97,6 +98,15 @@
+ *
+ * @run main/othervm -XX:+UnlockDiagnosticVMOptions -XX:+UnlockExperimentalVMOptions -Xmx1g -Xms1g
+ *      -XX:+UseShenandoahGC -XX:ShenandoahGCHeuristics=adaptive -XX:ShenandoahGCMode=generational
+ *      TestAllocIntArrays
```
@earthling-amzn will make a PR to address such changes as this across all the tests.

earthling-amzn replied:
https://bugs.openjdk.org/browse/JDK-8338695

earthling-amzn commented on test/hotspot/gtest/gc/shenandoah/test_shenandoahNumberSeq.cpp:
```diff
@@ -34,45 +34,130 @@
+  void print(HdrSeq& seq, const char* msg) {
+    std::cout << "[";
+    for (int i = 0; i <= 100; i += 10) {
+      std::cout << "\t" << seq.percentile(i);
```
@ysramakrishna will fix this.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.cpp:
```diff
@@ -382,33 +450,103 @@ void ShenandoahHeapRegion::print_on(outputStream* st) const {
+  size_t card_index = scanner->card_index_for_addr(start);
+  size_t num_cards = words / CardTable::card_size_in_words();
+
+  if (dirty_only) {
```
@earthling-amzn filed https://bugs.openjdk.org/browse/JDK-8338528

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahHeapRegion.cpp:
```diff
@@ -382,33 +450,103 @@ void ShenandoahHeapRegion::print_on(outputStream* st) const {
 
-void ShenandoahHeapRegion::oop_iterate_humongous(OopIterateClosure* blk) {
+// DO NOT CANCEL.  If this worker thread has accepted responsibility for scanning a particular range of addresses, it
+// must finish the work before it can be cancelled.
```
https://bugs.openjdk.org/browse/JDK-8338528

earthling-amzn commented on src/hotspot/share/gc/shared/gcConfiguration.cpp:
```diff
@@ -42,18 +43,22 @@ GCName GCConfiguration::young_collector() const {
 
+  if (UseShenandoahGC) {
+#if INCLUDE_SHENANDOAHGC
+    if (strcmp(ShenandoahGCMode, "generational") == 0) {
```
@earthling-amzn will fix this per https://bugs.openjdk.org/browse/JDK-8338620

rkennke replied:
ok

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahFullGC.cpp:
```diff
@@ -273,8 +290,11 @@ class ShenandoahPrepareForMarkClosure: public ShenandoahHeapRegionClosure {
   void heap_region_do(ShenandoahHeapRegion *r) {
-    _ctx->capture_top_at_mark_start(r);
-    r->clear_live_data();
+    // TODO: Add API to heap to skip free regions
```
@earthling-amzn will confirm that the regions are asserted to be in the correct state when they become affiliated.

rkennke replied:
ok

kdnilsen commented on test/hotspot/jtreg/gc/shenandoah/generational/TestOldGrowthTriggers.java:
```diff
@@ -0,0 +1,115 @@
+            array[replace_index] = array[replace_index].multiply(array[derive_index]);
+            break;
+        }
+        if ((i & 0x3) == 0x3) {
+        } else {
+        }
```
@kdnilsen to look at this.

shipilev commented on src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp:
```diff
@@ -599,29 +685,73 @@ void ShenandoahConcurrentGC::op_final_mark() {
       // Verify before arming for concurrent processing.
       // Otherwise, verification can trigger stack processing.
       if (ShenandoahVerify) {
         heap->verifier()->verify_during_evacuation();
```
I am surprised this verification step does not fail, because it is supposed to check that `has_forwarded = true` as the global GC state. The patch moves the switch for `has_forwarded` past this check. 

Edit: I know see the relaxation in verifier. Anyhow, I think this verification step is unnecessary: [JDK-8338662](https://bugs.openjdk.org/browse/JDK-8338662).

shipilev commented on src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp:
```diff
@@ -599,29 +685,73 @@ void ShenandoahConcurrentGC::op_final_mark() {
     heap->prepare_concurrent_roots();
 
-    if (!heap->collection_set()->is_empty()) {
+    if (!heap->collection_set()->is_empty() || has_in_place_promotions(heap)) {
+      // Even if the collection set is empty, we need to do evacuation if there are regions to be promoted in place.
+      // Concurrent evacuation takes responsibility for registering objects and setting the remembered set cards to dirty.
```
My initial reaction was that we are breaking away from the invariant that "evac means evac: cset is not empty, there are forwarded objects", and generational hijacks evacuation for its promotional needs. In some sense, this is still valid: promotion young->old can be logically treated as "evacuation". Was this the intent?

I suggest introducing a local var: `bool is_relocating_objects = !heap->collection_set()->is_empty()`, an use it in this code.

earthling-amzn replied:
@earthling-amzn will introduce local variable for readability.

kdnilsen replied:
Tracking this in https://bugs.openjdk.org/browse/JDK-8339615

shipilev commented on src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp:
```diff
@@ -718,8 +861,14 @@ void ShenandoahEvacUpdateCleanupOopStorageRootsClosure::do_oop(oop* p) {
-      ShenandoahHeap::atomic_clear_oop(p, obj);
+      shenandoah_assert_generations_reconciled();
+      if (_heap->is_in_active_generation(obj)) {
+        // Here we are asserting that an unmarked from-space object is 'correct'. There seems to be a legitimate
+        // use-case for accessing from-space objects during concurrent class unloading. In all modes of Shenandoah,
+        // concurrent class unloading only happens during a global collection.
+        shenandoah_assert_correct(p, obj);
```
I noticed this before and removed that assert here: https://github.com/openjdk/jdk/commit/b9d49dcef22ab81a087d890bbac0329a5244a2ef#diff-3876d66a8561c2fe5826c6217603b76efde78847ec46b11936af52158f58c27cR721

shipilev commented on src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp:
```diff
@@ -922,7 +1074,9 @@ void ShenandoahConcurrentGC::op_init_updaterefs() {
   heap->prepare_update_heap_references(true /*concurrent*/);
   heap->set_update_refs_in_progress(true);
-
+  if (ShenandoahVerify) {
+    heap->verifier()->verify_before_updaterefs();
+  }
```
This looks to be JDK-8336914, should disappear after the merge.

shipilev reviewed (commented):
Another short review round:

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp:
```diff
@@ -536,10 +592,32 @@ void ShenandoahConcurrentGC::op_init_mark() {
+      // The current implementation of swap_remembered_set() copies the write-card-table
+      // to the read-card-table. The remembered sets are also swapped for GLOBAL collections
+      // so that the verifier works with the correct copy of the card table when verifying.
+      // TODO: This path should not really depend on ShenandoahVerify.
```
@earthling-amzn will take a look at this.

earthling-amzn replied:
https://bugs.openjdk.org/browse/JDK-8338763

earthling-amzn commented on test/hotspot/jtreg/gc/shenandoah/TestSmallHeap.java:
```diff
@@ -23,7 +24,7 @@
 
 /*
- * @test
+ * @test default
```
https://bugs.openjdk.org/browse/JDK-8338695

earthling-amzn commented on test/hotspot/jtreg/gc/shenandoah/generational/TestSimpleGenerational.java:
```diff
@@ -0,0 +1,125 @@
+ *      -XX:+UseShenandoahGC -XX:ShenandoahGCMode=generational
+ *      gc.shenandoah.generational.TestSimpleGenerational
+ */
+public class TestSimpleGenerational {
```
https://bugs.openjdk.org/browse/JDK-8338695

earthling-amzn commented on test/hotspot/jtreg/gc/shenandoah/TestShenandoahRegionLogging.java:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2017, 2018, Red Hat, Inc. All rights reserved.
+ * Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
```
I'm not sure how to fix this. `TestParallelRefproceSanity.java` was deleted here: `2785875de3bf86ab20779c5a24ec7e0f69e452aa` and `TestShenandoahRegionLogging.java` was independently added first as `TestShenandoahLogRotation.java` in `2bc25de76aa6e8434c2b0ccfe9ac8a86f21e38cb`. It seems git has just decided the content of these files is similar enough to be a rename.

rkennke replied:
For the next PR (you close this, and open a new one, as soon as we are satisfied with first rounds of reviews), squash everything in a single commit, rather then importing the complete history of GenShen into the PR. Then you can also get the renaming correct, I think.

earthling-amzn replied:
Okay, will try that.

ysramakrishna commented on src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.inline.hpp:
```diff
@@ -234,7 +249,11 @@ inline oop ShenandoahBarrierSet::AccessBarrier<decorators, BarrierSetT>::oop_loa
+  shenandoah_assert_marked_if(nullptr, value,
+                              !CompressedOops::is_null(value) &&
+                              ShenandoahHeap::heap()->is_evacuation_in_progress() &&
+                              !(ShenandoahHeap::heap()->active_generation()->is_young() &&
+                              ShenandoahHeap::heap()->heap_region_containing(value)->is_old()));
```
Done: https://github.com/openjdk/shenandoah/pull/478

Can resolve.

ysramakrishna commented on src/hotspot/share/gc/shenandoah/heuristics/shenandoahAdaptiveHeuristics.cpp:
```diff
@@ -21,16 +22,21 @@
  * questions.
  *
  */
-
```
@ysramakrishna : https://bugs.openjdk.org/browse/JDK-8338780

rkennke replied:
ok

ysramakrishna commented on src/hotspot/share/gc/shenandoah/heuristics/shenandoahAdaptiveHeuristics.cpp:
```diff
@@ -21,16 +22,21 @@
-
 #include "precompiled.hpp"
 
+
```
@ysramakrishna : https://bugs.openjdk.org/browse/JDK-8338780

rkennke commented:
I left a few replies. Consider all other open discussions from me as 'resolved'.

earthling-amzn commented on src/hotspot/share/gc/shenandoah/shenandoahAsserts.cpp:
```diff
@@ -71,6 +71,9 @@ void ShenandoahAsserts::print_obj(ShenandoahMessageBuffer& msg, oop obj) {
   msg.append("    %3s marked strong\n",              ctx->is_marked_strong(obj) ? "" : "not");
   msg.append("    %3s marked weak\n",                ctx->is_marked_weak(obj) ? "" : "not");
   msg.append("    %3s in collection set\n",          heap->in_collection_set(obj) ? "" : "not");
+  if (heap->mode()->is_generational() && !obj->is_forwarded()) {
```
We can show the address of the forwardee here, instead of nothing.

shipilev replied:
OK

shipilev commented on src/hotspot/cpu/aarch64/gc/shenandoah/shenandoahBarrierSetAssembler_aarch64.cpp:
```diff
@@ -387,18 +416,13 @@ void ShenandoahBarrierSetAssembler::store_at(MacroAssembler* masm, DecoratorSet
-  if (val == noreg) {
-    BarrierSetAssembler::store_at(masm, decorators, type, Address(tmp3, 0), noreg, noreg, noreg, noreg);
-  } else {
-    // Barrier needs uncompressed oop for region cross check.
```
This seems to be upstreamable separately from GenShen.

earthling-amzn replied:
I agree, but I'd like to try and upstream it through this PR.

shipilev commented on src/hotspot/cpu/x86/gc/shenandoah/shenandoahBarrierSetAssembler_x86.cpp:
```diff
@@ -787,6 +891,63 @@ void ShenandoahBarrierSetAssembler::cmpxchg_oop(MacroAssembler* masm,
+  __ shrptr(end, CardTable::card_shift());
+  __ subptr(end, addr); // end --> cards count
+
+  __ mov64(tmp, disp);
```
Does it need to be `mov64`? I think `MacroAssembler::movptr(Address dst, intptr_t src)` would shorten the code if the base is 32-bit encodable.

earthling-amzn replied:
Will test this.

earthling-amzn replied:
We could change this, but it's already under `#ifdef _LP64` and there is a separate block of assembly for the 32-bit case. I don't think we need to change this.

shipilev commented on src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.cpp:
```diff
@@ -124,6 +132,13 @@ void ShenandoahBarrierSet::on_thread_detach(Thread *thread) {
       gclab->retire();
     }
 
+    if (ShenandoahCardBarrier) {
```
Fairly weird to see this conditionalized on `ShenandoahCardBarrier`. This seems to be a crude check for generational. But I think this is unnecessary, since PLAB would be null if generational is disabled?

earthling-amzn replied:
Much of the code here uses specific barrier flags like this (`ShenandoahStackWatermarkBarrier`, is a proxy for SATB mode, for example). But you are right, the PLAB should be null in non-generational modes and the accessor for the generational heap will assert if we try to use it here outside of non-generational modes.

earthling-amzn replied:
https://bugs.openjdk.org/browse/JDK-8340395

shipilev commented on src/hotspot/share/gc/shenandoah/shenandoahBarrierSet.inline.hpp:
```diff
@@ -323,16 +364,23 @@ template <typename T>
 template <class T, bool HAS_FWD, bool EVAC, bool ENQUEUE>
 void ShenandoahBarrierSet::arraycopy_work(T* src, size_t count) {
-  assert(HAS_FWD == _heap->has_forwarded_objects(), "Forwarded object status is sane");
+  // We allow forwarding in young generation and marking in old generation
+  // to happen simultaneously.
+  assert(_heap->mode()->is_generational() || HAS_FWD == _heap->has_forwarded_objects(), "Forwarded object status is sane");
```
This is my problem: `has_forwarded_objects` is the safety statement about the state of the heap. When it is true, we know we have to do fwdptr resolves for extra safety. The argument for skipping it should be _very strong_. So my question is then:

> There are never forwarded objects in old regions during a concurrent mark of the old generation.

Is that really guaranteed? To be more precise, what exactly guarantees that the objects _referenced by this array_ are in the same generation, and thus are not forwarded when we are doing the old marking? I can get how we can guarantee that the _array itself_ is not forwarded, but what about its elements?

earthling-amzn replied:
Old marking doesn't _really_ run concurrently with evacuations. Old marking will also never interrupt a young cycle (though old marking may be interrupted by young cycles). As I think on this more, I believe that comment is incorrect and we should be able to restore the original assert. Will test this.

earthling-amzn replied:
I tested this and the assert still failed, but for a reason we haven't considered yet. As we've discussed previously in this pull request, the generational mode may promote regions in place during the evacuation phase _without_ forwarding or actually moving any objects. Historically, the `gc_state` for evacuation and forwarding has been somewhat conflated. In these cases when the collector is _only_ promoting regions in place, this barrier is not necessary. I'll address this issue and other related concerns as part of: https://bugs.openjdk.org/browse/JDK-8339616.

earthling-amzn replied:
This assert has been restored to its original form.

shipilev commented on src/hotspot/share/gc/shenandoah/shenandoahConcurrentGC.cpp:
```diff
@@ -632,39 +735,47 @@ void ShenandoahConcurrentGC::op_final_mark() {
-  void work(uint worker_id) {
+  void work(uint worker_id) override {
+    if (GENERATIONAL) {
+      Thread* worker_thread = Thread::current();
+      ShenandoahThreadLocalData::enable_plab_promotions(worker_thread);
```
Just inline `worker_thread`.

shipilev commented on src/hotspot/share/gc/shenandoah/shenandoahControlThread.cpp:
```diff
@@ -371,6 +375,16 @@ void ShenandoahControlThread::request_gc(GCCause::Cause cause) {
 }
 
 void ShenandoahControlThread::handle_requested_gc(GCCause::Cause cause) {
+  // For normal requested GCs (System.gc) we want to block the caller. However,
+  // for whitebox requested GC, we want to initiate the GC and return immediately.
+  // The whitebox caller thread will arrange for itself to wait until the GC notifies
+  // it that has reached the requested breakpoint (phase in the GC).
+  if (cause == GCCause::_wb_breakpoint) {
+    _requested_gc_cause = cause;
+    _gc_requested.set();
+    return;
+  }
```
Does this relate to GenShen, or is it a general issue?

earthling-amzn replied:
Yes, I believe it _could_ also affect other modes. I'm not sure why we've never seen the failure fixed by this change upstream. I will prepare a separate upstream PR for this.

earthling-amzn replied:
https://bugs.openjdk.org/browse/JDK-8340400

kdnilsen commented:
rkennke writes:
> I'm finishing my day and send out today's round of reviews.
> I must say that in many places I find it difficult to convince myself that the changes in shared Shenandoah code don't affect singly-gen-Shenandoah. But at the same time, I am not sure what we can do about it. We certainly don't want to spend another year refactoring this, and risk to introduce new bugs. I suspect that you guys did a lot of due diligence to get this correct. I'd welcome your perspective on this matter.

We exercise all CI tests and soak tests, including multiple performance-critical service workloads, on both single-generation and multi-generation modes of Shenandoah.  We have tracked performance on both. Many of our changes to shared code have already been integrated into single-generation Shenandoah independent of the GenShen integration.  The remaining "entanglement" of changes that affect shared code have been carefully scrutinized trhough multiple reviews.

earthling-amzn commented:
We are closing this pull request and reopening another without all the commit history and addressed feedback. Please refer to https://github.com/openjdk/jdk/pull/21273.
